Create Box1: experimental design (#191)
jaybee84/ml-in-rd: content/box_1_experimental_design.md

# Understanding experimental design of machine learning to inform requirements for data

> Component of ML experiments

Reviewer: ```suggestion
### Component of ML experiments

```

Machine learning algorithms identify patterns that explain or fit a given dataset.

> So every machine learning algorithm (_model_) expects the following fundamental parts as input: 1. a _training dataset_ , 2. a _cross-validation dataset_ , 3. a _held-out validation dataset_ .

Reviewer: I would argue that this is not true, i.e., the model doesn't really care about cross-validation or held-out validation sets but we sure do as experimentalists!

Author: :) fair point

Author: ```suggestion
So every machine learning algorithm (_model_) should be provided with the following fundamental parts as input: 1. a _training dataset_ , 2. a _cross-validation dataset_ , 3. a _held-out validation dataset_ .
```

Typically each of these input datasets consists of a matrix (or dataframe or table) where features form the rows and samples form the columns.
A _training dataset_ is used to expose the model to underlying patterns among the features present in the data of interest.

> A _cross-validation dataset_ is a smaller test dataset which is used during training of the model to iteratively update its parameters (i.e. _hyperparameter_tuning_ or _model tuning_).

Reviewer: ```suggestion
A _cross-validation dataset_ is a smaller test dataset which is used during training of the model to iteratively update its parameters (i.e., _hyperparameter_tuning_ or _model tuning_).
```

A _held-out validation set_ is a new or unseen test dataset that is used to test whether the patterns learned by the model can be generalized to new data points.

> While the cross-validation dataset helps the model iteratively update its parameters to learn important patterns in the training data, the held-out validation set helps test the generalizability of the model.

Reviewer: Hm.. I think this discussion of cross-validation sets coming before you break down what cross-validation is potentially confusing. 

Could you limit the discussion to training and testing (and other terms people use for testing/validation sets) in the first section here?

Author: yes.. will do this in my revision

Generalizability of a model is its ability to recognize patterns that can help predict the class or an outcome for previously unseen data.
High generalizability of a model on previously unseen data suggest that the model has identified fundamental patterns in the data that may also inform our knowledge regarding the question of interest for which the experiment was designed.
Generalizability can be affected if _data leakage_ occurs during training of the model, i.e. if a model sees the same or similar data points in both the training set and the held-out test set.

> Ensuring absence of any overlap or relatedness among data points or samples (e.g. samples with familial relationship, samples from same patient) used in the training set and held-out test set is important to avoid _data leakage_ during model training.

Reviewer: ```suggestion
Generalizability can be affected if _data leakage_ occurs during training of the model, i.e., if a model sees the same or similar data points in both the training set and the held-out test set.
Ensuring absence of any overlap or relatedness among data points or samples (e.g., samples with familial relationship, samples from same patient) used in the training set and held-out test set is important to avoid _data leakage_ during model training.
```

Specifically in cases of rare genetic diseases where many samples can contain familial relationships, special care should be taken while crafting the training and testing sets to ensure that no data leakage occurs and the trained model has high generalizability.

> Training and testing

Reviewer: ```suggestion
### Training and testing

```

> The implementation of a machine learning experiment begins with splitting a single dataset of interest such that 70% of the data is used as the _training dataset_, 20% of the data is used as the _cross-validation dataset_, and remaining 10% of the data is used as the _hold-out validation dataset_.

Reviewer: Checking my understanding of the set up you would like to get across here – there's a 70% training set that gets split up into k-folds for hyperparameter tuning. Tuned models are then tested on the 20% cross-validation set to select the "final model" (i.e., the model with the best performance on the cross-validation set). The performance of the final model is tested on the 10% hold-out set. Is that correct?

Author: Yes, your interpretation here is correct. But I am aware of other studies that had ~90% data as training set > broken into k-folds for hyperparameter tuning+cross validation > test on a ~10% holdout set. 

In your experience have you come across other variations? 
Also, I wonder if we should describe the second version,  instead/in addition to the one I mention in the text. What do you think?

Author: May be I can resolve this confusion if I refer to the different inputs like : 
1. Training dataset (used for hyper-parameter tuning)
2. Evaluation dataset (used for model testing using cross-validation)
3. Prediction dataset (held-out dataset used for final prediction and model performance)

Model testing happens with both 2 and 3 which seems to be the root of this confusion .. I think..

Reviewer: I made this comment after reading the _whole paragraph_ – this + what is line 27 is what confused me.  I feel like the following:

>  ~90% data as training set > broken into k-folds for hyperparameter tuning+cross validation > test on a ~10% holdout set.

Might be easier to explain to someone who is being introduced to the concepts for the first time? But what you outline:

>   1. Training dataset (used for hyper-parameter tuning)
>   2. Evaluation dataset (used for model testing using cross-validation)
>   3. Prediction dataset (held-out dataset used for final prediction and model performance)

Is considerably more clear in my opinion. I think the main takeaway from this exchange, though, is that we should only describe **one approach.** I also expect that the visualization might help either way.

This makes sure that all the datasets involved in training and testing a model maintain uniformity in the features.
In case of rare diseases where multiple datasets may be combined to make a large enough training dataset, special care is taken to standardize the features and the patterns therein.
The iterative training stage helps the model learn important patterns in the training dataset and then use the cross-validation dataset to test for errors in prediction and update its learning parameters (_hyperparameter tuning_).
Multiple approaches exist for cross-validation e.g. leave-p-out cross-validation, leave-one-out cross-validation, k-fold cross-validation, Monte-Carlo random subsampling cross-validation [@doi:10.1007/978-1-4614-6849-3].
In case of k-fold cross-validation, a given dataset is shuffled randomly and split into _k_ parts.

> One of the _k_ parts is reserved as the _cross-validation dataset_ and the rest are used for training.

Reviewer: I find this confusing with the 20% of the data is used as the cross-validation set in the first sentence.

Author: Yeah.. I think this is where I mixed up 2 different approaches while writing.. 

In the first sentence, I am referring to approach 1 where the cross validation set is separate from the k-fold set used for hyperparameter tuning. In this one, I have referred to approach 2 where cross-validation + hyper-parameter tuning is done using the same training set broken down into k-folds. At first k-folds used for grid-search during the tuning stage and then k-folds used for cross-validation during model-fitting stage.

Any thoughts about one or the other approach?

# Understanding experimental design of machine learning to inform requirements for data

### Components of ML experiments

Machine learning algorithms identify patterns that explain or fit a given dataset.
Every machine learning algorithm (_model_) goes through a _training_ phase where it identifies underlying patterns in a given dataset, and a _testing_ phase where the model applies the identified patterns to unseen data points.
So every model should be provided with the following fundamental parts as input: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_ .

> Typically each of these input datasets consists of a matrix (or dataframe or table) where features form the rows and samples form the columns.

Reviewer: Other way around for most applications no?

Author: yes.. you are correct... I think I was visualizing a gene expression heatmap while writing this...

Author: ```suggestion
Typically each of these input datasets consists of a matrix (or dataframe or table) where samples form the rows and features form the columns.
```
